### PR TITLE
Daml lf numeric

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR include the engine changes for Numerics (see #2289).
When necessary old decimal are map to `(Numeric 10)`

Other runtime components (such as codegen, extractor, navigator) still assume there is only decimal.

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala [hunk 1]
```diff
@@ -4,35 +4,240 @@
...
+
+    def buildPrimType(primType: DamlLf1.PrimType, args: DamlLf1.Type*) =
+      DamlLf1.Type
+        .newBuilder()
+        .setPrim(DamlLf1.Type.Prim.newBuilder().setPrim(primType).addAllArgs(args.asJava))
+        .build()
 
+    val TDecimal = TNumeric(Ast.TNat(10))
+
+    "transparently apply TNat(10) to TNumeric if version =< 1.dev" in {
```
[the remark this answers is not in the document]
Comment (gerolf-da) [reply daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala]: Add comment that this test doesn't type check but only translates potentially nonsensical protobuf to an AST, that doesn't necessarily type check.

File: daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala [hunk 2]
```diff
@@ -4,35 +4,240 @@
...
+  "decodeExpr" should {
+
+    def toProto(b: BuiltinInfo) =
+      DamlLf1.Expr.newBuilder().setBuiltin(b.protoName).build()
+
+    def toScala(b: BuiltinInfo) =
+      Ast.EBuiltin(b.lfName)
+
+    val numericBuilttins = DecodeV1.builtinInfos.filter(_.decimalLegacy)
+    val nonNumericBuiltins =
```
Comment (gerolf-da) [inline daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala]: Add comment that we don't need to check with all non numeric builtins.

File: daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala [hunk 3]
```diff
@@ -4,35 +4,240 @@
...
+      DecodeV1.builtinInfos.filter(b =>
+        b.protoName == DamlLf1.BuiltinFunction.ADD_INT64 || b.protoName == DamlLf1.BuiltinFunction.APPEND_TEXT)
+    assert(nonNumericBuiltins.length == 2)
+
+    "transparently apply TNat(10) to Numeric builtins if version =< 1.dev" in {
```
Comment (gerolf-da) [inline daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala]: Add `FixMe` for release.

File: daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala [hunk 4]
```diff
@@ -4,35 +4,240 @@
...
+        }
+
+        forEvery(negativeTetsCases) { builtinInfo =>
+          decoder.decodeExpr(toProto(builtinInfo)) shouldBe toScala(builtinInfo)
+        }
+      }
+
+    }
+
+    "translate TNumeric such as if version > 1.6" in {
```
Comment (gerolf-da) [inline daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala]: `as is`.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Decimal.scala [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  val MaxValue: BigDecimal = BigDecimal("9999999999999999999999999999.9999999999")
+  val MinValue: BigDecimal = -MaxValue
+
+  final def fromBigDecimal(x0: BigDecimal): Either[String, BigDecimal] =
+    Numeric.fromBigDecimal(10, x0)
+
+  final def assertFromBigDecimal(x: BigDecimal): BigDecimal =
+    assertRight(fromBigDecimal(x))
+
+  private val hasExpectedFormat =
```
Comment (gerolf-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Decimal.scala]: Rename to `expectedFormat`

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Numeric.scala [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+import scala.math.BigDecimal
+import scala.util.Try
+
+object Numeric {
+
+  val legacyDecimalScale = 10
+  val maxPrecision = 38
+
+  // double check that the BigDecimal has the expected scale and precision
```
Comment (gerolf-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Numeric.scala]: Change the comment as discussed offline.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Numeric.scala [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+  def divide(scale: Int, x: BigDecimal, y: BigDecimal): Either[String, BigDecimal] = {
+    assertNumeric(scale, x)
+    assertNumeric(scale, y)
+    checkForOverflow(scale, x.bigDecimal.divide(y.bigDecimal, scale, ROUND_HALF_EVEN))
+  }
+
+  def toLong(scale: Int, x: BigDecimal): Either[String, Long] = {
+    assertNumeric(scale, x)
+    Try(x.bigDecimal.setScale(0, ROUND_DOWN).longValueExact()).toEither.left.map(
```
Comment (gerolf-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Numeric.scala]: Double check behavior `ROUND_DOWN` on negative numbers.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Numeric.scala [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+      _ => s"(Numeric $scale) $x does not fit into an Int64",
+    )
+  }
+
+  final def round(scale: Int, newScale: Long, x: BigDecimal): Either[String, BigDecimal] = {
```
Comment (gerolf-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Numeric.scala]: Exhaustively test these functions.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Numeric.scala [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    x compareTo y
+  }
+
+  final def fromBigDecimal(scale: Int, x: BigDecimal): Either[String, BigDecimal] =
+    checkScale(scale, x).flatMap(checkForOverflow(scale, _))
+
+  def assertFromBigDecimal(scale: Int, x: BigDecimal): BigDecimal =
+    assertRight(fromBigDecimal(scale, x))
+
+  private val ExpectedUnscaledFormat = """[-+]?(\d{1,38})(\.\d{0,38})?""".r
```
Comment (gerolf-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Numeric.scala]: Deduplicate the regex between `Numeric.scala` and `DecodeV1.scala`.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Numeric.scala [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+  def assertUnscaledFromString(s: String): BigDecimal =
+    assertRight(unscaledFromString(s))
+
+  def toString(x: BigDecimal): String = {
+    val s = x.bigDecimal.stripTrailingZeros.toPlainString
+    if (s.contains(".")) s else s + ".0"
+  }
+
+  private val ExpectedScaledFormat = """-?([1-9]\d*|0).(\d*)?""".r
```
Comment (gerolf-da) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Numeric.scala]: Should be the same as in the decoder.

Comment (gerolf-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Numeric.scala]: Drop the `?` and check for null of the decimal part when using the regex.

**Comment by remyhaemmerle-da** [comment]
Those changes have or will be push in smaller pieces.

**Pull request closed without merging**